client/test/error-telemetry.test.js

```diff
@@ -0,0 +1,48 @@
+jest.mock('fs');
+
+const assert         = require('assert');
+const fs             = require('fs');
+const ErrorTelemetry = require('../lib/error-telemetry');
+const mkdirp         = require('mkdirp');
+const logger = require('winston');
+
+describe('Error Telemetry Logging', () => {
+  beforeEach(() => {
+    /* Make sure memfs is flushed every time */
+    fs.volume.reset();
+  });
+
+  describe('setup() call', () => {
+    it('should not return a Promise', () => {
+      const response = (new ErrorTelemetry()).setup();
+      assert(response instanceof Promise);
+    });
+
+    // FIXME: only hackish, the end goal is definitely not to forward to another file
+    it('writing to essentials logging file should forward to another', () => {
+      jest.useFakeTimers();
+
+      // Write before setup to make sure the file is already present
+      const logsDir = '/evergreen/jenkins/var/logs/';
+      mkdirp.sync(logsDir);
+      fs.writeFileSync(logsDir + 'essentials.log.0', '{"timestamp":1523451065975,"level":"SEVERE","message":"WAT"}');
+
+      const response = (new ErrorTelemetry()).setup();
+      assert(response instanceof Promise);
+
+      // Write again
+      fs.appendFileSync(logsDir + 'essentials.log.0', '{"timestamp":1523451065975,"level":"SEVERE","message":"WAT2"}');
+
+      assert(!fs.existsSync('/tmp/test'));
+
+      setTimeout(function() {
+        logger.error('YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA');
+        assert(fs.existsSync('/tmp/test'));
+        assert.equal('MESSAGE =WAT\nMESSAGE=WAT2', fs.readFileSync('/tmp/test','utf8'));
+      }, 5000);
+      jest.runAllTimers();
```

Comment:
Because of this call, the timeout is executed directly and so you fall in a concurrency problem I think. Your `ErrorTelemetry` method does not have the time to do its stuff, especially the Tail that seems to use an interval meaning it's potentially not real time.

Option 1) you can edit the code and place something like the following to know if it's really a concurrency problem or not:
```
// in line 22, replace the arrow function by
(done) => {

// instead of lines 38-43
while(!fs.existsSync('/tmp/test');
assert.equal('MESSAGE=WAT\nMESSAGE=WAT2', fs.readFileSync('/tmp/test','utf8'));
done();
```
With that, the test will wait the file exists and only at that point it will check the content. If the file will never exist, the test will timeout itself. The `done` tells the script runner that the test is asynchronous and it must wait until the method is called.

Option 2) do not use the jest mocked timers to have a real 5sec wait.

Reply:
@Wadeck thanks for the help.

I cannot make any of those work. If I try what you propose with the loop above, it seems to be blocked and never return.

I suspect this is due to the Javascript threading model and using an active loop is just keeping it looping, but nothing else can move forward while we're looping (?)